File: Section06OptionalChallenges/DivideAndConquer_countZeros.js
function countZeroes(arr) {
  if (!arr || arr.length <= 0) {
    return 0;
  }
  let index = Math.trunc(arr.length / 2);
  let zeroCountDone = false;
  while (!zeroCountDone) {
    if (index === 0 || arr.length === index + 1) {
      if (arr.length === index + 1) {
        index++;
      }
      zeroCountDone = true;
      break;
    } else {
      if (arr[index] === 0) {
        if (index > 0 && arr[index - 1] === 1) {
          zeroCountDone = true;
          break;
        } else if (index === 1) {
          index = 0;
        } else {
          index = Math.max(0, Math.trunc((arr.length - index) / 2));
        }
      } else {
        index = Math.min(arr.length - 1, Math.trunc((arr.length + index) / 2));
      }
    }
  }
  return arr.length - index;
}

console.log(countZeroes([1,1,1,1,0,0]));
console.log(countZeroes([1,0,0,0,0]));
console.log(countZeroes([0,0,0]));
console.log(countZeroes([1,1,1,1]));

File: Section06OptionalChallenges/Section06OptionalChallenges_DOTNET/DivideAndConquer_CountZeros.cs
using System;

namespace Section06OptionalChallenges.DivideAndConquer;

public partial class DivideAndConquer
{
  public static int CountZeros(int[] arr)
  {
    if (arr == null || arr.Length <= 0)
    {
      return 0;
    }
    var index = (int)Math.Floor((double)arr.Length / 2.0);
    var zerCountDone = false;
    while (!zerCountDone)
    {
      if (index == 0 || arr.Length - 1 == index)
      {
        if (arr.Length - 1 == index)
        {
          index++;
        }
        zerCountDone = true;
        break;
      }
      else
      {
        if (arr[index] == 0)
        {
          if (index > 0 && arr[index - 1] == 1)
          {
            zerCountDone = true;
            break;
          }
          else if (index == 1)
          {
            index = 0;
          }
          else
          {
            index = Math.Max(0, (int)Math.Floor(((double)arr.Length - index) / 2.0d));
          }
        }
        else
        {
          index = Math.Min(arr.Length - 1, (int)Math.Floor(((double)arr.Length + index) / 2.0d));
        }
      }
    }
    return arr.Length - index;
  }
}